scylla/src/transport/load_balancing/token_and_latency_aware.rs

```diff
@@ -0,0 +1,278 @@
+use super::{LatencyAwarePolicy, LoadBalancingPolicy, Plan, Statement, TokenAwarePolicy};
+use crate::transport::cluster::ClusterData;
+use std::time::Duration;
+
+/// A token- and latency-aware load balancing policy.
+#[derive(Debug)]
+pub struct TokenAndLatencyAwarePolicy {
+    inner: TokenAwarePolicy,
+}
+
+impl TokenAndLatencyAwarePolicy {
+    pub fn new(
+        exclusion_threshold: f64,
+        retry_period: Duration,
+        update_rate: Duration,
+        minimum_measurements: usize,
+    ) -> Self {
+        Self {
+            inner: TokenAwarePolicy::new(Box::new(LatencyAwarePolicy::new(
+                exclusion_threshold,
+                retry_period,
+                update_rate,
+                minimum_measurements,
+            ))),
+        }
+    }
+}
+
+impl Default for TokenAndLatencyAwarePolicy {
+    fn default() -> Self {
+        Self {
+            inner: TokenAwarePolicy::new(Box::new(LatencyAwarePolicy::default())),
```

Comment:
I think the hierarchy should be"
`LatencyAwarePolicy` -> `TokenAwarePolicy` -> `RoundRobin`

`TokenAwarePolicy` would choose replicas and shuffle them with round robin, then LatencyAware would use this plan, but filter out nodes that have high latencies.

This doesn't really fit the `ChildLoadBalancingPolicy` trait, but we could make `LatencyAwarePolicy` keep `inner: Box<dyn LoadBalancingPolicy>` inside of it.

Reply:
> * `LatencyAwarePolicy` will only be applied to replicas, after trying all the replicas other nodes will be tried without regard for latency

I think this is what we want. Ultimately we're going to contact replicas anyway, so we should prefer to do it directly. There are cases where contacting a non-replica coordinator is better because replicas are overloaded, but in my opinion it's far more likely that contacting a replica is more beneficial in the average case.

> * There is no load balancing among replicas. Each query with a token for replica `A` will always hit `A` first, then `B` and `C`. With round robin the load is spread equally among the replicas, because the first replica changes.

Could you clarify your concerns? Assuming replication factor == 3, each token would determine a set of 3 replicas that own it. It sounds correct to apply latency awareness on this filtered list of candidates.

Also, take into account that we should generally assume that gathering latency information is a costly operation (it's prohibitively costly right now with a mutex, but it will still be costly on atomics). Therefore, getting this information from a pre-filtered list of replicas sounds more efficient than gathering latency information on all nodes (imagine we have tens of nodes in multiple datacenters) and only then applying token awareness. I think the current approach is valid.

Reply:
My main problem with this implementation is lack of load balancing among the replicas. There is no round robin, so for a given token we will always try replicas in the same order. This will make the first replica handle all the requests, while the others stay idle.
I understand that this is a `LatencyTokenPolicy`, not `LatencyTokenRoundRobinPolicy`, but I think that we should introduce a policy that uses the latency information and is usable by production code.

Reply:
> I think this is what we want. Ultimately we're going to contact replicas anyway, so we should prefer to do it directly. There are cases where contacting a non-replica coordinator is better because replicas are overloaded, but in my opinion it's far more likely that contacting a replica is more beneficial in the average 

You're right. Still after trying the replicas it would be good to try other nodes in the order of their latencies, but I'm not sure if there's an easy way to do it.